### PR TITLE
Builder-Webpack5: Categorize builder error

### LINKED ISSUE #24030: categorize webpack builder errors

### PR DESCRIPTION
**Pull request by yannbf**
Closes #24030

## What I did

Webpack builder errors before were just an object with a `compilation` property.
This PR revamps both the `start` and `build` processes in the Webpack5 builder, adding categorized Webpack errors, turning Webpack compilation errors into a single, composite error.

Turns out there was a bug since Storybook 7 where, if you have Storybook successfully running on dev mode, but then introduce a compilation error, e.g. change the code in `Button.js` to have incorrect syntax like `conzt foo` instead of `const foo`, the error would not be propagated to the terminal anymore. This was the case in Storybook 6.5, and it stopped working because of changes in the webpack-dev-middleware library. This PR brings that behavior back.

#### The changes in this PR are covered in the following automated tests:
- [ ] stories
- [x] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

_This section is mandatory for all contributions. If you believe no manual test is necessary, please state so explicitly. Thanks!_

Please check unit tests

### Documentation

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

This pull request has been released as version [`0.0.0-pr-24031-sha-553789e8`](https://npmjs.com/package/@storybook/cli/v/0.0.0-pr-24031-sha-553789e8). Install it by pinning all your Storybook dependencies to that version.
<details>
<summary>More information</summary>

| | |
| --- | --- |
| **Published version** | [`0.0.0-pr-24031-sha-553789e8`](https://npmjs.com/package/@storybook/cli/v/0.0.0-pr-24031-sha-553789e8) |
| **Triggered by** | @yannbf |
| **Repository** | [storybookjs/storybook](https://github.com/storybookjs/storybook) |
| **Branch** | [`yann/webpack-builder-error`](https://github.com/storybookjs/storybook/tree/yann/webpack-builder-error) |
| **Commit** | [`553789e8`](https://github.com/storybookjs/storybook/commit/553789e83af83c57179659c26223d87f043f8460) |
| **Datetime** | Sun Sep  3 21:15:01 UTC 2023 (`1693775701`) |
| **Workflow run** | [6066789759](https://github.com/storybookjs/storybook/actions/runs/6066789759) |

To request a new release of this pull request, mention the `@storybookjs/core` team.

_core team members can create a new canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=24031`_
</details>

### CODE REVIEW

File: code/lib/core-events/src/errors/server-errors.test.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    data.error = new Error();
+    data.error.stats = {
+      compilation: {
+        errors: compilationErrors,
+      },
+    };
+
+    const webpackError = new WebpackCompilationError(data);
+
+    const expectedErrorMessage = compilationErrors
```
Comment (shilman) [inline code/lib/core-events/src/errors/server-errors.test.ts:42]: Can you hard code this instead? Hard to read what the result is supposed to be in the current form

**Review by shilman — APPROVED** [review]
Nice one! 💪

File: code/lib/core-events/src/errors/server-errors.test.ts (hunk 1 above)
Comment (yannbf) [reply code/lib/core-events/src/errors/server-errors.test.ts]: Done!

File: code/lib/core-events/src/errors/server-errors.test.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+    expect(webpackError.message).toBe(error.message);
+  });
+
+  it('should correctly handle error with error within error', () => {
+    const error = new Error('Custom error message');
+    const data = {
+      error: new Error(),
+    };
+    data.error = error;
```
Comment (tmeasday) [inline code/lib/core-events/src/errors/server-errors.test.ts:21]: Is this supposed to be `data.error.error = error`?

File: code/lib/core-events/src/errors/server-errors.ts [hunk 3]
```diff
@@ -137,3 +137,48 @@ export class InvalidStoriesEntryError extends StorybookError {
...
+          this.errorMessage += `${e.name}: ${e.message}\n\n`;
+        });
+      } else {
+        this.errorMessage = data.error.message;
+      }
+    } else if (data.error.compilation?.errors) {
+      data.error.compilation.errors.forEach((e: Error) => {
+        this.errorMessage += `${e.name}: ${e.message}\n\n`;
+      });
+    }
```
Comment (tmeasday) [inline code/lib/core-events/src/errors/server-errors.ts:178]: We don't have a test for this bit?

Comment (yannbf) [reply code/lib/core-events/src/errors/server-errors.ts:178]: great catch! done!

File: code/lib/core-events/src/errors/server-errors.test.ts (hunk 2 above)
Comment (yannbf) [reply code/lib/core-events/src/errors/server-errors.test.ts]: great catch! fixed

File: code/builders/builder-webpack5/src/index.ts [hunk 4]
```diff
@@ -195,8 +196,7 @@ const starter: StarterFunction = async function* starterGeneratorFn({
   }
 
   if (stats.hasErrors()) {
-    // eslint-disable-next-line @typescript-eslint/no-throw-literal
-    throw stats;
+    throw new WebpackCompilationError({ error: stats });
```
Comment (ndelangen) [inline code/builders/builder-webpack5/src/index.ts:199]: Don't merge this suggested code, it's just a starting point for the convo:

It seems weird to me that we'd lump these different types of issues together.
Should we separate these into multiple different  
```suggestion
    throw new WebpackCompilationStatsError({ stats });
```

Or we separate the different input types like so:
```suggestion
    throw new WebpackCompilationError({ stats });
```

WDTY?

**Review by tmeasday — APPROVED** [review]
LGTM!

**Pull request merged**